File: tools/imaging_on_duty_validator.py
#!/usr/bin/env python3
"""
Imaging On-Duty Validator — pdfplumber cross-check for MISC Duty Rota.
SCOPED TO: Imaging On-Duty (radiology_duty) ONLY.

Usage:
    from imaging_on_duty_validator import validate_with_pdfplumber
    corrected = validate_with_pdfplumber(pdf_path, current_extraction)

Or standalone:
    python3 imaging_on_duty_validator.py <pdf_path>
"""

import pdfplumber
import re
import json
import sys
import os

# Section-locked doctors
SECTION_LOCKS = {
    'a. dhafiri': 'MSK',
    'ahmed al dhafiri': 'MSK',
    'fatimah albahhar': 'MSK',
    'hassan ghafouri': 'MSK',
}

INVALID_FRAGMENTS = {'F.', 'A.', 'N.', 'H.', 'M.', 'S.', 'K.', 'R.', 'T.', 'E.',
                     'Al', 'Al-', 'Dr.', 'Dr'}


def _is_imaging_duty_pdf(pdf_path):
    """Detect if PDF is an Imaging On-Duty rota."""
    name = os.path.basename(pdf_path).upper()
    return any(k in name for k in ['MISC', 'IMAGING', 'WEEKLY DUTY ROTA', 'MISC DUTY'])


def validate_with_pdfplumber(pdf_path, current_extraction):
    """
    Cross-check and correct Imaging On-Duty extraction using pdfplumber.
    Returns corrected data in the SAME format as input.
    """
    if not _is_imaging_duty_pdf(pdf_path):
        return current_extraction

    corrections = []

    try:
        # Import the extractor for ground-truth comparison
        sys.path.insert(0, os.path.dirname(__file__))
        from rota_extractor import extract_rota
        ground_truth = extract_rota(pdf_path)
    except Exception as e:
        print(f"[IMAGING VALIDATOR] pdfplumber extraction failed: {e}")
        return current_extraction

    # Build lookup from ground truth: (date, shift, section) → record
    gt_lookup = {}
    for r in ground_truth:
        key = (r['date'], r['shift'], r['section'])
        gt_lookup[key] = r

    corrected = []
    for record in current_extraction:
        r = dict(record)

        # 1. Section boundary check
        name_lower = (r.get('name', '') or '').lower()
        section = (r.get('section', '') or '').upper()
        for doc_name, locked_section in SECTION_LOCKS.items():
            if doc_name in name_lower and section != locked_section.upper():
                corrections.append({
                    'type': 'section_leak',
                    'name': r.get('name'),
                    'found_in': section,
                    'belongs_to': locked_section,
                })
                r['name'] = ''

        # 2. Name fragment cleanup
        name = (r.get('name', '') or '').strip()
        if name in INVALID_FRAGMENTS:
            corrections.append({
                'type': 'name_fragment',
                'fragment': name,
            })
            r['name'] = ''

        # 3. Phone format check
        phone = (r.get('phone', '') or '').replace('-', '').replace(' ', '')
        if phone and not re.match(r'^05\d{8}$', phone):
            corrections.append({
                'type': 'bad_phone',
                'name': r.get('name'),
                'phone': r.get('phone'),
            })
            r['phone'] = ''
            r['phoneUncertain'] = True

        if (r.get('name', '') or '').strip():
            corrected.append(r)

    if corrections:
        print(f"\n[IMAGING VALIDATOR] {len(corrections)} correction(s):")
        for c in corrections[:10]:
            print(f"  {c['type']}: {c}")

    return corrected


if __name__ == '__main__':
    pdf_path = sys.argv[1] if len(sys.argv) > 1 else None
    if not pdf_path:
        print("Usage: python3 imaging_on_duty_validator.py <pdf_path>")
        sys.exit(1)

    # Simulate: run both extractors and compare
    sys.path.insert(0, os.path.dirname(__file__))
    from rota_extractor import extract_rota

    print("=== Running pdfplumber extraction ===")
    records = extract_rota(pdf_path)

    print(f"\n=== Validation Results ===")
    print(f"Records extracted: {len(records)}")
    sections = {r['section'] for r in records}
    for s in sorted(sections):
        mods = {r['modality'] for r in records if r['section'] == s}
        for m in mods:
            cnt = sum(1 for r in records if r['section'] == s and r['modality'] == m)
            print(f"  {s} / {m}: {cnt}")

    # Check role exclusivity
    violations = 0
    for r in records:
        all_names = r['consultant'] + r['residents'] + r['fellow'] + r['assistant_associate']
        if len(all_names) != len(set(all_names)):
            violations += 1
            dups = [n for n in all_names if all_names.count(n) > 1]
            print(f"  ROLE DUP: {r['date']} {r['shift']} {r['section']}: {set(dups)}")

    # Check section locks
    for r in records:
        for role in ['consultant', 'residents', 'fellow', 'assistant_associate']:
            for name in r[role]:
                for doc, locked in SECTION_LOCKS.items():
                    if doc in name.lower() and r['section'] != locked:
                        print(f"  SECTION LEAK: {name} in {r['section']} (should be {locked})")

    print(f"\nRole violations: {violations}")
    print(f"Phone coverage: {sum(1 for r in records for p in r['phones'].values() if p)}/{sum(len(r['phones']) for r in records)}")

File: assets/js/parsers/imaging_on_duty_validator.js
// ═══════════════════════════════════════════════════════
// imaging_on_duty_validator.js — Post-extraction accuracy layer
// SCOPED TO: Imaging On-Duty (radiology_duty) ONLY
// ═══════════════════════════════════════════════════════
// This validator runs AFTER parseRadiologyDutyPdfEntries returns.
// It cross-checks the extraction for known structural errors:
//   1. Section boundary leaks (MSK doctor in THORACIC, etc.)
//   2. Role misassignment (fellow promoted to consultant)
//   3. Name fragment cleanup ("F." standalone, "A." without last name)
//   4. Duplicate names across roles in same record
//   5. Phone number confidence check
//
// If any correction is made, it's logged to the console for audit.
// If any check fails gracefully, the original record is returned unchanged.
// ═══════════════════════════════════════════════════════

/**
 * Section-locked doctors. Each doctor belongs to ONE section only.
 * If they appear in the wrong section's entries → remove them.
 */
const SECTION_LOCKS = {
  'A. Dhafiri': 'MSK',
  'Ahmed Al Dhafiri': 'MSK',
  'Dr. Ahmed Al Dhafiri': 'MSK',
  'Fatimah Albahhar': 'MSK',
  'Hassan Ghafouri': 'MSK',
};

/**
 * Known name fragments that should never appear as standalone names.
 * Usually caused by PDF.js splitting "F. Alkhabaz" across AM/PM boundary.
 */
const INVALID_NAME_FRAGMENTS = new Set([
  'F.', 'A.', 'N.', 'H.', 'M.', 'S.', 'K.', 'R.', 'T.', 'E.',
  'Al', 'Al-', 'Dr.', 'Dr',
]);

/**
 * Validate and correct a set of Imaging On-Duty extraction records.
 * ONLY call this for radiology_duty entries.
 *
 * @param {Array} records - parsed entries from parseRadiologyDutyPdfEntries
 * @param {string} sourceText - raw PDF text (for context, not re-parsed)
 * @returns {Array} corrected records in the SAME format
 */
function validateImagingOnDutyExtraction(records, sourceText) {
  if (!Array.isArray(records) || !records.length) return records;

  // Only apply to imaging duty — quick sanity check
  const hasImagingSection = records.some(r =>
    (r.section || '').match(/NEURO|BODY|THORAC|MSK|BREAST|PEDIA/i)
  );
  if (!hasImagingSection) return records; // not imaging duty, skip

  let corrections = 0;
  const corrected = records.map(record => {
    const r = { ...record };
    const section = (r.section || '').toUpperCase();

    // ── CHECK 1: Section boundary enforcement ──────────────────
    for (const [doctorName, lockedSection] of Object.entries(SECTION_LOCKS)) {
      const nameNorm = doctorName.toLowerCase();
      if (section === lockedSection.toUpperCase()) continue; // correct section, skip

      // Check if this doctor appears in any role for the WRONG section
      for (const role of ['name']) {
        if ((r[role] || '').toLowerCase().includes(nameNorm)) {
          console.log(`[IMAGING VALIDATOR] Section leak: "${doctorName}" belongs to ${lockedSection}, found in ${section}. Removing.`);
          r[role] = '';
          corrections++;
        }
      }
    }

    // ── CHECK 2: Name fragment cleanup ─────────────────────────
    const name = (r.name || '').trim();
    if (INVALID_NAME_FRAGMENTS.has(name)) {
      console.log(`[IMAGING VALIDATOR] Name fragment: "${name}" is not a valid name. Clearing.`);
      r.name = '';
      corrections++;
    }

    // ── CHECK 3: Role assignment — fellow vs consultant ────────
    // If a name appears as both Fellow and Consultant in entries
    // for the same date+shift+section, keep only the role that matches
    // the PDF's structural row (consultant row vs fellow sub-row).
    // This is enforced at the group level below.

    return r;
  });

  // ── CHECK 4: Cross-record role deduplication ──────────────
  // Group entries by date+shift+section, then check for duplicate names
  const groups = {};
  corrected.forEach((r, idx) => {
    const key = `${r.date || ''}|${r.startTime || ''}|${r.section || ''}`;
    if (!groups[key]) groups[key] = [];
    groups[key].push({ record: r, idx });
  });

  for (const [key, entries] of Object.entries(groups)) {
    const nameToRole = {};
    entries.forEach(({ record }) => {
      const name = (record.name || '').toLowerCase();
      const role = (record.role || '').toLowerCase();
      if (!name) return;
      if (nameToRole[name] && nameToRole[name] !== role) {
        // Same name, different roles in same shift/section
        // Keep the one that's more specific (resident > fellow > consultant)
        const priority = { 'consultant': 3, 'fellow': 2, 'resident': 1, 'assistant/associate consultant': 0 };
        const existingPri = priority[nameToRole[name]] || 99;
        const newPri = priority[role] || 99;
        if (newPri > existingPri) {
          // Current is higher priority (consultant) — remove the duplicate lower-priority entry
          const dup = entries.find(e =>
            (e.record.name || '').toLowerCase() === name
            && (e.record.role || '').toLowerCase() === nameToRole[name]
          );
          if (dup) {
            console.log(`[IMAGING VALIDATOR] Role dedup: "${record.name}" in both "${nameToRole[name]}" and "${role}". Keeping "${role}".`);
            corrected[dup.idx] = { ...dup.record, name: '' };
            corrections++;
          }
        }
      }
      nameToRole[name] = role;
    });
  }

  // ── CHECK 5: Phone number format validation ──────────────
  corrected.forEach(r => {
    const phone = (r.phone || '').replace(/[^0-9]/g, '');
    if (phone && !phone.match(/^05\d{8}$/)) {
      console.log(`[IMAGING VALIDATOR] Bad phone format: "${r.phone}" for "${r.name}". Clearing.`);
      r.phone = '';
      r.phoneUncertain = true;
      corrections++;
    }
  });

  // Remove entries with empty names (cleaned by checks above)
  const final = corrected.filter(r => (r.name || '').trim().length > 1);

  if (corrections > 0) {
    console.log(`[IMAGING VALIDATOR] ${corrections} correction(s) applied. ${records.length} → ${final.length} entries.`);
  }

  return final;
}
